mmv1/third_party/terraform/services/bigtable/resource_bigtable_instance.go

```
@@ -793,6 +825,12 @@ func resourceBigtableInstanceClusterReorderTypeListFunc(diff tpgresource.Terrafo
 				return fmt.Errorf("Error setting cluster diff: %s", err)
 			}
 		}
+
+		// Do NOT ForceNew for node scaling factor. Return the error on diffs.
+		oNsf, nNsf := diff.GetChange(fmt.Sprintf("cluster.%d.node_scaling_factor", i))
+		if oNsf != nNsf {
+			return fmt.Errorf("Immutable field 'node_scaling_factor' cannot be updated.")
+		}
```

Instead of returning errors, using `ForeceNew` is a standard way to handle immutable field.

---

We explicitly don't want to ForceNew in this case, as we don't want folks to accidentally wipe out their database by changing a field. I haven't used Terraform personally, so I don't know what the actual user experience is in this case, e.g. would they ever get a prompt to warn them about recreating the instance, or would it silently proceed and make this change?

I know the other fields are implemented with ForceNew, but I know the SME on one of those fields isn't fond of that ForceNew behavior either.

Ron initially suggested that we omit this code entirely if the server side will cause a failure. The grpc API will fail if called in this case. However, we have a quirk with our Go client where an update mask prevents the field from getting updated at all. In this case, Terraform will try the update, not change the field, and then the verification will fail (it will query the state and give the original attribute).

Given all this, I'd like to still validate that the attribute doesn't change in Terraform. Is this acceptable? Happy to discuss on chat or GVC if necessary

---

> We explicitly don't want to ForceNew in this case, as we don't want folks to accidentally wipe out their database by changing a field. I haven't used Terraform personally, so I don't know what the actual user experience is in this case, e.g. would they ever get a prompt to warn them about recreating the instance, or would it silently proceed and make this change?
> 
The terraform will not silently make the change.
The terraform plan will show that the resource will be replaced and the changed immutable field will be shown with `# forces replacement`. 

```
  # google_compute_instance.test-vm-1 must be replaced
-/+ resource "google_compute_instance" "test-vm-1" {
         ......
          ~ source                     = "test-vm-10111701" -> "test" # forces replacement
        ......
```

> I know the other fields are implemented with ForceNew, but I know the SME on one of those fields isn't fond of that ForceNew behavior either.
> 
Do you know what fields don't have `ForceNew: true`?

> Ron initially suggested that we omit this code entirely if the server side will cause a failure. The grpc API will fail if called in this case. However, we have a quirk with our Go client where an update mask prevents the field from getting updated at all. In this case, Terraform will try the update, not change the field, and then the verification will fail (it will query the state and give the original attribute).
> 
> Given all this, I'd like to still validate that the attribute doesn't change in Terraform. Is this acceptable? Happy to discuss on chat or GVC if necessary
I will discuss with our team and get back to you soon.

---

I will check on if other fields aren't covered by "ForceNew", and also check w/ our wider team on the stance of using ForceNew in general

---

The only field not covered by ForceNew is ClusterId. Though that's somewhat of a primary key field anyway. If I try that in Terraform, TF lets the update go through, but the update fails in the server (which is expected)

I'll share some notes offline on our team's stance on this

---

Per offline discussion, we have implemented this via ForceNew